src/python/pants/engine/build_files.py

```diff
@@ -114,17 +96,16 @@ def _raise_did_you_mean(address_family, name):
                      .format(name, address_family.namespace, possibilities))
 
 
-@rule(UnhydratedStruct,
-      [Select(AddressMapper),
-       SelectProjection(AddressFamily, Dir, 'spec_path', Address),
-       Select(Address)])
-def resolve_unhydrated_struct(address_mapper, address_family, address):
+@rule(UnhydratedStruct, [Select(AddressMapper), Select(Address)])
+def resolve_unhydrated_struct(address_mapper, address):
   """Given an Address and its AddressFamily, resolve an UnhydratedStruct.
 
   Recursively collects any embedded addressables within the Struct, but will not walk into a
   dependencies field, since those are requested explicitly by tasks using SelectDependencies.
   """
 
+  address_family = yield Get(AddressFamily, Dir(address.spec_path))
```

Comment:
If no `Dir -> AddressFamily` path exists in the rule graph, does it error at this line? Would the previous use of `SelectProjection` have errored out earlier, during the rule graph construction?

Reply:
Yep, exactly. And that's the primary downside of this change as it stands: we would lose the validation that all rules are satisfiable.

Have been thinking more about this though, and I think that with some very, very basic parsing (essentially, "capture the RHS of yield statements"), we could get back the static checking... which might be justified.

Reply:
...and also, as it stands the exception would be created in the rust code rather than literally on this line. It looks like it is easy to send an exception to the generator though with `generator.throw()`.

Reply:
I think "capture the RHS of yield statements" is a highly reasonable goal, but I don't know if it needs to be done by parsing -- that being said I remember python's ast module is supposed to be not awful? I can't see what the exact code would look like right now, but since we know we _can_ get the desired info we want before invoking the python rule definition (with the previous example using `SelectProjection`), I think there's hopefully a middle ground where we can specify rule graphs without too much ceremony (probably exactly the same in spirit as parsing for yield statements). I would love to parse some python for yield statements if it gets us there, but I think there might be a neater way -- will think on it.

Reply:
This is now  resolved: all `Get` statements in the body of an `@rule` are statically checked at `RuleGraph` construction time.